esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -2298,6 +2264,66 @@ bool CFileSprayEx::onSprayVariable(IEspContext &context, IEspSprayVariable &req,
     return true;
 }
 
+void CFileSprayEx::checkDZScopeAccessAndSetSpraySourceDFUFileSpec(IEspContext &context, const char *srcPlane, const char *srcHost,
+    const char *srcFile, MemoryBuffer &srcXML, IDFUfileSpec *srcDFUfileSpec)
+{
+    if(srcXML.length() == 0)
+    {
+        RemoteMultiFilename rmfn;
+        SocketEndpoint ep(srcHost);
+        if (ep.isNull())
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Cannot resolve source network IP from %s.", srcHost);
+
+        rmfn.setEp(ep);
+        StringBuffer fnamebuf(srcFile);
+        fnamebuf.trim();
+        rmfn.append(fnamebuf.str());    // handles comma separated files
+        srcDFUfileSpec->setMultiFilename(rmfn);
+
+        ForEachItemIn(i, rmfn)
+        {
+            StringBuffer path;
+            rmfn.item(i).getLocalPath(path);
+
+            //A spray source could be: '/path/*'.
+            StringBuffer dirPath, dirTail;
+            splitFilename(path, &dirPath, &dirPath, &dirTail, &dirTail);
+
+            const char* s = streq(dirTail.str(), "*") ? dirPath : path;
+            SecAccessFlags permission = getDropZoneScopePermissions(context, srcPlane, s, srcHost);
+            if (permission < SecAccess_Read)
+                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone Scope %s %s not allowed for user %s (permission:%s). Read Access Required.",
+                    isEmptyString(srcPlane) ? srcHost : srcPlane, s, context.queryUserId(), getSecAccessFlagName(permission));
+        }
+    }
+    else
+    {
+        srcXML.append('\0');
```

Comment:
bug: should not write to what was a read-only const MemoryBuffer , but was cast. The casts should be removed too..

There's no need either, there's a form of createPTreeFromXMLString that takes a length that could be used (with a modified setFromXML).

Reply:
question: out of interest, what are the use cases / code paths - that pass in srcXML like this to these services?

Reply:
I haven't heard any use cases / code paths. I traced the code inside the setFromXML() and found that what is needed inside the xml. Then, I created my test xml.